README.md
# MobPush For Flutter

# 注意,该仓库是基于MobPush官方插件进行的优化,支持了Dart空安全和修复了几个bug

这是一个基于  MobPush 功能的扩展的 Flutter 插件。使用此插件能够帮助您在使用 Flutter 开发应用时，快速地实现推送功能。

插件主页：https://pub.dartlang.org/packages/mobpush_plugin

Demo例子：https://github.com/MobClub/MobPush-for-Flutter/mobpush_plugin

>注意：老版本插件mobpush([https://pub.dartlang.org/packages/mobpush](https://pub.dartlang.org/packages/mobpush))不再更新，请使用当前插件mobpush_plugin([https://pub.dartlang.org/packages/mobpush_plugin](https://pub.dartlang.org/packages/mobpush_plugin))
>

开始集成

1.参考 Flutter 官方插件集成文档

在pubspec.yaml文件中加入下面依赖

```
dependencies:
  mobpush_plugin:
```
然后执行：flutter packages get 导入package
在你的dart工程文件中，导入下面头文件，开始使用

```
import 'package:mobpush_plugin/mobpush_plugin.dart';
```

**iOS：**

平台配置参考[ iOS 集成文档](http://wiki.mob.com/mobpush-for-ios/)

实现 1. ：获取 appKey 和 appSecret
实现 5.1：配置 appkey 和 appSecret

**Android：**

导入 MobPush 相关依赖

在项目根目录的build.gradle中添加以下代码：

```
dependencies {
    classpath 'com.android.tools.build:gradle:3.2.1'
    classpath 'com.mob.sdk:MobSDK:+'
    classpath 'com.google.gms:google-services:4.0.1' //不需要FCM厂商推送无需配置
}
```
在 /android/app/build.gradle 中添加以下代码：

```
apply plugin: 'com.android.application'
apply from: "$flutterRoot/packages/flutter_tools/gradle/flutter.gradle"
// 导入MobSDK
apply plugin: 'com.mob.sdk'

MobSDK {
    appKey "您的Mob平台appKey"
    appSecret "您的Mob平台appSecret"

	//配置MobPush
    MobPush {
    	//配置厂商推送（可选配置，不需要厂商推送可不配置，需要哪些厂商推送只需配置哪些厂商配置即可）
        devInfo {
        	//配置小米厂商推送
            XIAOMI {
                appId "您的小米平台appId"
                appKey "您的小米平台appKey"
            }

			//配置华为厂商推送
            HUAWEI {
                appId "您的华为平台appId"
            }

			//配置魅族厂商推送
            MEIZU {
                appId "您的魅族平台appId"
                appKey "您的魅族平台appKey"
            }

			//配置FCM厂商推送
            FCM {
                //设置默认推送通知显示图标
                iconRes "@mipmap/default_ic_launcher"
            }

            //配置OPPO厂商推送
            OPPO {
                appKey "您的OPPO平台appKey"
                appSecret "您的OPPO平台appSecret"
            }

            //配置VIVO厂商推送
            VIVO {
                appId "您的VIVO平台appId"
                appKey "您的VIVO平台appKey"
            }
        }
    }
}
```
3.  在MainActivity的onCreate中添加以下代码：


```
 @Override
  protected void onCreate(Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);
    GeneratedPluginRegistrant.registerWith(this);
  }
```
在项目的/android/app的AndroidManifest.xml文件中添加：

```
<application
        android:name=".CustomApplication"
        ... 
        >
        ...
</application>
```
 
**接口方法说明**

（1）设置隐私协议授权状态 ：
updatePrivacyPermissionStatus

```
MobpushPlugin.updatePrivacyPermissionStatus(true);
```


（2）设置远程推送环境，向用户授权（仅 iOS）：

setCustomNotification

```
if (Platform.isIOS) {
      MobpushPlugin.setCustomNotification();
}
```

（3）设置远程推送环境  (仅 iOS)：

setAPNsForProduction

```
if (Platform.isIOS) {
     // 开发环境 false, 线上环境 true
      MobpushPlugin.setAPNsForProduction(false)
}
```
（4）添加推送回调监听（接收自定义透传消息回调、接收通知消息回调、接收点击通知消息回调）

addPushReceiver

```
MobpushPlugin.addPushReceiver(_onEvent, _onError);

void _onEvent(Object event) {

}

void _onError(Object event) {
    
}
```

（5）停止推送

stopPush

```
MobpushPlugin.stopPush();
```

（6）重新打开推送服务

restartPush

```
MobpushPlugin.restartPush();
```
（7）是否已停止接收推送

isPushStopped

```
MobpushPlugin.isPushStopped();
```

（8）设置别名

setAlias

```
MobpushPlugin.setAlias("别名").then((Map<String, dynamic> aliasMap){
	String res = aliasMap['res'];
	String error = aliasMap['error'];
	String errorCode = aliasMap['errorCode'];
	print(">>>>>>>>>>>>>>>>>>>>>>>>>>> setAlias -> res: $res error: $error");
});
```

（9）获取别名

getAlias

```
MobpushPlugin.getAlias().then((Map<String, dynamic> aliasMap){
	String res = aliasMap['res'];
	String error = aliasMap['error'];
	print(">>>>>>>>>>>>>>>>>>>>>>>>>>> getAlias -> res: $res error: $error");
});
```

(10）删除别名

deleteAlias

```
MobpushPlugin.deleteAlias("别名").then((Map<String, dynamic> aliasMap){
	String res = aliasMap['res'];
	String error = aliasMap['error'];
	print(">>>>>>>>>>>>>>>>>>>>>>>>>>> deleteAlias -> res: $res error: $error");
});
```

（11）添加标签

addTags

```
List tags = new List();
tags.add("tag1");
tags.add("tag2");
MobpushPlugin.addTags(tags).then((Map<String, dynamic> tagsMap){
	String res = tagsMap['res'];
	String error = tagsMap['error'];
	print(">>>>>>>>>>>>>>>>>>>>>>>>>>> addTags -> res: $res error: $error");
});
```

（12）获取标签

getTags

```
MobpushPlugin.getTags().then((Map<String, dynamic> tagsMap){
	List<String> resList =  List<String>.from(tagsMap['res']);
	String error = tagsMap['error'];
	print(">>>>>>>>>>>>>>>>>>>>>>>>>>> getTags -> res: $resList error: $error");
});;
```

（13）删除标签

deleteTags

```
List tags = new List();
tags.add("tag1");
tags.add("tag2");
MobpushPlugin.deleteTags(tags).then((Map<String, dynamic> tagsMap){
	String res = tagsMap['res'];
	String error = tagsMap['error'];
	print(">>>>>>>>>>>>>>>>>>>>>>>>>>> deleteTags -> res: $res error: $error");
});
```

（14）清空标签

cleanTags

```
MobpushPlugin.cleanTags().then((Map<String, dynamic> tagsMap){
	String res = tagsMap['res'];
	String error = tagsMap['error'];
	print(">>>>>>>>>>>>>>>>>>>>>>>>>>> cleanTags -> res: $res error: $error");
});
```

（15）发送本地通知

addLocalNotification

```
MobpushPlugin.addLocalNotification();
```

（16）绑定手机号

bindPhoneNum

```
MobpushPlugin.bindPhoneNum("110");
```

（17）测试模拟推送，用于测试

send

```
/**
    * 测试模拟推送，用于测试
    * type：模拟消息类型，1、通知测试；2、内推测试；3、定时
    * content：模拟发送内容，500字节以内，UTF-8
    * space：仅对定时消息有效，单位分钟，默认1分钟
    * extras: 附加数据，json字符串
    */
MobpushPlugin.send(int type, String content, int space, String extras).then((Map<String, dynamic> sendMap){
	String res = sendMap['res'];
	String error = sendMap['error'];
	print(">>>>>>>>>>>>>>>>>>>>>>>>>>> send -> res: $res error: $error");
});
```

（18）设置点击通知是否跳转默认页 (仅andorid)

setClickNotificationToLaunchMainActivity

```
MobpushPlugin.setClickNotificationToLaunchMainActivity (bool enable);
```

（19）移除本地通知(仅andorid)

removeLocalNotification

```
MobpushPlugin.removeLocalNotification(int notificationId);
```

（20）清空本地通知(仅andorid)

clearLocalNotifications

```
MobpushPlugin.clearLocalNotifications();
```

（21）设置通知栏icon，不设置默认取应用icon(仅andorid)

setNotifyIcon

```
MobpushPlugin.setNotifyIcon(String resId);
```

（22）设置应用在前台时是否隐藏通知不进行显示，不设置默认不隐藏通知(仅andorid)

setAppForegroundHiddenNotification

```
MobpushPlugin.setAppForegroundHiddenNotification(bool hidden);
```

（23）设置通知静音时段（推送选项）(仅andorid)

setSilenceTime

```
/**
   * 设置通知静音时段（推送选项）(仅andorid)
   * @param startHour   开始时间[0~23] (小时)
   * @param startMinute 开始时间[0~59]（分钟）
   * @param endHour     结束时间[0~23]（小时）
   * @param endMinute   结束时间[0~59]（分钟）
   */
MobpushPlugin.setSilenceTime(int startHour, int startMinute, int endHour, int endMinute)
```

（24）设置角标 (仅 iOS)

setBadge


```
MobpushPlugin.setBadge(int badge);
```

（25）清空角标，不清除通知栏消息记录 (仅 iOS)

clearBadge

```
MobpushPlugin.clearBadge();
```


（26）获取注册Id

getRegistrationId

```
MobpushPlugin.getRegistrationId().then((Map<String, dynamic> ridMap) {
	print(ridMap);
	String regId = ridMap['res'].toString();
	print('------>#### registrationId: ' + regId);
});

```

CHANGELOG.md
## 0.0.1

* TODO: Describe initial release.

## 1.0.1

* Support for mobpush.

## 1.0.2

* Support all API for MobPush.

## 1.0.3

* Add How to Use Demo for MobPush.

## 1.0.4

* 1、完善所有iOS API；
* 2、提供完整Demo示例；

## 1.0.5

* 1、iOS 新增设置前台通知展示类型（Badge、Sound、Alert）;
* 2、更新Demo其他API界面为列表展示;

## 1.0.6

* 1、调整普通接口通过result回调;
* 2、接口回调改为Map类型用于提供接口错误信息回调;

## 1.0.7

* 1、iOS 消息点击回调 bug fixed. 

## 1.0.9

* 1、BUG 修复

## 1.1.2

* 1.优化体验
* 2.demo厂商更新
* 3.修复已知BUG

## 1.1.3

* 1.优化体验
* 2.demo厂商更新
* 3.修复已知BUG

## 1.1.4

* 1.bug修复

## 1.1.5

* 1.优化体验
* 2.demo更新华为厂商

## 1.1.6

* 1.优化别名获取

## 2.0.1
* 1.在官方插件的基础上,支持了Dart空安全
* 2.修复了android端,native和flutter通信时,没有在主线程造成的bug

example/test/widget_test.dart
// This is a basic Flutter widget test.
//
// To perform an interaction with a widget in your test, use the WidgetTester
// utility that Flutter provides. For example, you can send tap and scroll
// gestures. You can also use WidgetTester to find child widgets in the widget
// tree, read text, and verify that the values of widget properties are correct.

import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

import 'package:mobpush_plugin_example/main.dart';

void main() {
  testWidgets('Verify Platform version', (WidgetTester tester) async {
    // Build our app and trigger a frame.
    await tester.pumpWidget(MyApp());

    // Verify that platform version is retrieved.
    expect(
      find.byWidgetPredicate(
        (Widget widget) =>
            widget is Text && widget.data!.startsWith('Running on:'),
      ),
      findsOneWidget,
    );
  });
}


example/lib/main.dart
import 'dart:convert';
import 'dart:io';

import 'package:flutter/material.dart';
import 'dart:async';

import 'package:flutter/services.dart';
import 'package:mobpush_plugin/mobpush_plugin.dart';
import 'package:mobpush_plugin/mobpush_custom_message.dart';
import 'package:mobpush_plugin/mobpush_notify_message.dart';
import 'app_notify_page.dart';
import 'click_container.dart';
import 'local_notify_page.dart';
import 'notify_page.dart';
import 'other_api_page.dart';
import 'timing_notify_page.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return new MaterialApp(
      home: MainApp(),
    );
  }
}

class MainApp extends StatefulWidget {
  @override
  _MainAppState createState() {
    return _MainAppState();
  }
}

class _MainAppState extends State<MainApp> {
  String _sdkVersion = 'Unknown';
  String _registrationId = 'Unknown';

  @override
  void initState() {
    super.initState();

    initPlatformState();

    if (Platform.isIOS) {
      MobpushPlugin.setCustomNotification();
      MobpushPlugin.setAPNsForProduction(false);
    }
    MobpushPlugin.addPushReceiver(_onEvent, _onError);

    //上传隐私协议许可
    MobpushPlugin.updatePrivacyPermissionStatus(true);
  }

  void _onEvent(Object? event) {
    print('>>>>>>>>>>>>>>>>>>>>>>>>>>>onEvent:' + event.toString());
    setState(() {
      Map<String, dynamic> eventMap = json.decode(event.toString());
      Map<String, dynamic> result = eventMap['result'];
      int action = eventMap['action'];

      switch (action) {
        case 0:
          MobPushCustomMessage message =
              new MobPushCustomMessage.fromJson(result);
          showDialog(
            context: context,
            builder: (context) {
              return AlertDialog(
                content: Text(message.content),
                actions: <Widget>[
                  TextButton(
                    onPressed: () {
                      Navigator.pop(context);
                    },
                    child: Text("确定"),
                  )
                ],
              );
            },
          );
          break;
        case 1:
          MobPushNotifyMessage message =
              new MobPushNotifyMessage.fromJson(result);
          break;
        case 2:
          MobPushNotifyMessage message =
              new MobPushNotifyMessage.fromJson(result);
          break;
      }
    });
  }

  void _onError(Object? event) {
    setState(() {
      print('>>>>>>>>>>>>>>>>>>>>>>>>>>>onError:' + event.toString());
    });
  }

  void _onAppNotifyPageTap() {
    setState(() {
      Navigator.push(context,
          new MaterialPageRoute(builder: (context) => new AppNotifyPage()));
    });
  }

  void _onNotifyPageTap() {
    setState(() {
      Navigator.push(context,
          new MaterialPageRoute(builder: (context) => new NotifyPage()));
    });
  }

  void _onTimingNotifyPageTap() {
    setState(() {
      Navigator.push(context,
          new MaterialPageRoute(builder: (context) => new TimingNotifyPage()));
    });
  }

  void _onLocalNotifyPageTap() {
    setState(() {
      Navigator.push(context,
          new MaterialPageRoute(builder: (context) => new LocalNotifyPage()));
    });
  }

  void _onOtherAPITap() {
    setState(() {
      Navigator.push(context,
          new MaterialPageRoute(builder: (context) => new OtherApiPage()));
    });
  }

  Future<void> initPlatformState() async {
    String sdkVersion;

    try {
      sdkVersion = await MobpushPlugin.getSDKVersion();
    } on PlatformException {
      sdkVersion = 'Failed to get platform version.';
    }
    try {
      MobpushPlugin.getRegistrationId().then((Map<String, dynamic> ridMap) {
        print(ridMap);
        setState(() {
          _registrationId = ridMap['res'].toString();
          print('------>#### registrationId: ' + _registrationId);
        });
      });
    } on PlatformException {
      _registrationId = 'Failed to get registrationId.';
    }

    // If the widget was removed from the tree while the asynchronous platform
    // message was in flight, we want to discard the reply rather than calling
    // setState to update our non-existent appearance.
    if (!mounted) return;

    setState(() {
      _sdkVersion = sdkVersion;
    });
  }

  // 复制到剪切板
  void _onCopyButtonClicked() {
    // 写入剪切板
    Clipboard.setData(ClipboardData(text: _registrationId));
    // 验证是否写入成功
    Clipboard.getData(Clipboard.kTextPlain).then((data) {
      String? text = data!.text;
      print('------>#### copyed registrationId: $text');
      if (text == _registrationId) {
        showDialog(
            context: context,
            builder: (context) {
              return AlertDialog(
                title: Text("恭喜🎉"),
                content: Container(
                  margin: EdgeInsets.only(top: 10, bottom: 30),
                  child: Text('复制成功！'),
                ),
                actions: <Widget>[
                  new FlatButton(
                    child: new Text("OK"),
                    onPressed: () {
                      Navigator.pop(context);
                    },
                  )
                ],
              );
            });
      }
    });
  }

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        appBar: AppBar(
          title: const Text('MobPushPlugin Demo'),
        ),
        body: Column(
          children: <Widget>[
            Expanded(
              child: Row(
                children: <Widget>[
                  Expanded(
                    child: ClickContainer(
                      content: 'App内推送',
                      res: 'assets/images/ic_item_app_nitify.png',
                      left: 15.0,
                      top: 15.0,
                      right: 7.5,
                      bottom: 7.5,
                      onTap: _onAppNotifyPageTap,
                    ),
                  ),
                  Expanded(
                    child: ClickContainer(
                      content: '通知',
                      res: 'assets/images/ic_item_notify.png',
                      left: 7.5,
                      top: 15.0,
                      right: 15.0,
                      bottom: 7.5,
                      onTap: _onNotifyPageTap,
                    ),
                  ),
                ],
              ),
            ),
            Expanded(
              child: Row(
                children: <Widget>[
                  Expanded(
                    child: ClickContainer(
                      content: '定时推送',
                      res: 'assets/images/ic_item_timing.png',
                      left: 15.0,
                      top: 7.5,
                      right: 7.5,
                      bottom: 7.5,
                      onTap: _onTimingNotifyPageTap,
                    ),
                  ),
                  Expanded(
                    child: ClickContainer(
                      content: '本地通知',
                      res: 'assets/images/ic_item_local.png',
                      left: 7.5,
                      top: 7.5,
                      right: 15.0,
                      bottom: 7.5,
                      onTap: _onLocalNotifyPageTap,
                    ),
                  ),
                ],
              ),
            ),
            Expanded(
              child: Row(
                children: <Widget>[
                  Expanded(
                    child: ClickContainer(
                      content: '其他API接口',
                      res: 'assets/images/ic_item_media.png',
                      left: 15.0,
                      top: 7.5,
                      right: 15.0,
                      bottom: 7.5,
                      onTap: _onOtherAPITap,
                    ),
                  ),
                ],
              ),
            ),
            Container(
              margin: EdgeInsets.only(left: 15.0, right: 15.0, bottom: 15.0),
              height: 60,
              child: Row(
                crossAxisAlignment: CrossAxisAlignment.start,
                children: <Widget>[
                  Text(
                    'SDK Version: $_sdkVersion\nRegistrationId: $_registrationId',
                    style: TextStyle(fontSize: 12),
                  ),
                  RaisedButton(
                    child: Text('复制'),
                    onPressed: _onCopyButtonClicked,
                  ),
                ],
              ),
            ),
          ],
        ),
      ),
    );
  }
}


lib/mobpush_notify_message.dart
/*
 * MobPush通知消息实体
 */
class MobPushNotifyMessage {
  // 共用属性

  /*
   * 消息标题
   */
  String? title;

  /*
   * 消息内容
   */
  String? content;

  /*
   * 消息时间戳
   */
  int? timestamp;

  //iOS 特有属性

  /*
   * 副标题
   */
  String? subTitle;
  
  /*
   * 消息时间戳
   */
  String? sound;

  /*
   * 角标
   */
  int? badge;

  //Android 特有属性

  /*
   * 大段文本和大图模式的样式内容
   */
  String? styleContent;

  /*
   * 消息Id
   */
  String? messageId;

  /*
   * 收件箱样式的内容
   */
  List<String>? inboxStyleContent;

  /*
   * 通知样式 android:(0：普通样式；1：大段文本；2：大图样式；3：收件箱样式)
   */
  int? style;

  /*
   * 推送通道：android:(0：mobPush；1：华为；2：小米；3：魅族；4：FCM；5：vivo)
   */
  int? channel;

  /*
   * 消息附加数据
   */
  Map<String, dynamic>? extrasMap;

  /*
   * 通知提示音
   */
  bool? voice;

  /*
   * 通知震动
   */
  bool? shake;

  /*
   * 通知呼吸灯
   */
  bool? light;

  MobPushNotifyMessage({
      this.title,
      this.content,
      this.messageId,
      this.timestamp,
      this.style,
      this.channel,
      this.voice,
      this.shake,
      this.extrasMap,
      this.inboxStyleContent,
      this.styleContent,
      this.light,
      this.badge,
      this.sound,
      this.subTitle});

  // MobPushNotifyMessage(this.title, this.content, this.timestamp, this.badge, this.sound, this.subTitle);

  factory MobPushNotifyMessage.fromJson(Map<String, dynamic> json) {
    return MobPushNotifyMessage(
        title: json['title'],
        content: json['content'],
        messageId: json['messageId'],
        timestamp: json['timestamp'],
        style: json['style'],
        channel: json['channel'],
        voice: json['voice'],
        shake: json['shake'],
        extrasMap: json['extrasMap'],
        inboxStyleContent: json['inboxStyleContent'],
        styleContent: json['styleContent'],
        light: json['light'],
        badge: json['badge'],
        sound: json['sound'],
        subTitle: json['subTitle']);
  }
}


example/lib/click_container.dart

import 'package:flutter/material.dart';

class ClickContainer extends StatefulWidget {
  final double? left;
  final double? top;
  final double? right;
  final double? bottom;
  final String? res;
  final String? content;
  final GestureTapCallback? onTap;

  ClickContainer(
    {
      this.content, 
      this.res, 
      this.left, 
      this.top, 
      this.right, 
      this.bottom, 
      this.onTap
    }
  );

  @override
  _ClickContainerState createState() {
    return _ClickContainerState();
  }
}

class _ClickContainerState extends State<ClickContainer> {

  @override
  Widget build(BuildContext context) {
    return new GestureDetector(
      child: Container(
        margin: EdgeInsets.only(left:widget.left!, top: widget.top!,
          right: widget.right!, bottom:widget.bottom!),
        decoration: BoxDecoration(
          color: Color(0xFFebf2ff),
          borderRadius: new BorderRadius.all(new Radius.circular(8.0)),
        ),
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            Image.asset(widget.res!),
            Text(widget.content!)
          ],
        ),
      ),
      onTap: widget.onTap,
    );
  }
}


lib/mobpush_local_notification.dart
import 'mobpush_notify_message.dart';

/*
 * MobPush本地通知实体
 */
class MobPushLocalNotification extends MobPushNotifyMessage {
  /*
   * 本地通知id，仅 Android 属性
   */
  int? notificationId;

  MobPushLocalNotification(
      {this.notificationId = 0,
      title,
      content,
      messageId,
      inboxStyleContent,
      timestamp,
      style,
      channel,
      extrasMap,
      voice,
      shake,
      styleContent,
      light,
      badge,
      sound,
      subTitle})
      : super(
            title: title,
            content: content,
            messageId: messageId,
            timestamp: timestamp,
            style: style,
            channel: channel,
            voice: voice,
            shake: shake,
            extrasMap: extrasMap,
            inboxStyleContent: inboxStyleContent,
            styleContent: styleContent,
            light: light,
            badge: badge,
            sound: sound,
            subTitle: subTitle);

  // MobPushLocalNotification(title, content, timestamp, badge, sound, subTitle);

  Map<String, dynamic> toJson() => {
        'notificationId': notificationId,
        'title': title,
        'content': content,
        'messageId': messageId,
        'inboxStyleContent': inboxStyleContent,
        'timestamp': timestamp,
        'style': style,
        'channel': channel,
        'extrasMap': extrasMap,
        'voice': voice,
        'shake': shake,
        'styleContent': styleContent,
        'light': light,
        'badge': badge,
        'sound': sound,
        'subTitle': subTitle
      };
}


lib/mobpush_plugin.dart
import 'dart:async';
import 'dart:convert';
import 'package:flutter/services.dart';
import 'mobpush_local_notification.dart';

typedef void EventHandler(Object? event);

class MobpushPlugin {
  static const MethodChannel _channel = const MethodChannel('mob.com/mobpush_plugin');
  static EventChannel _channelReciever = const EventChannel('mobpush_receiver');

  static Future<String> get platformVersion async {
    final String version = await _channel.invokeMethod('getPlatformVersion');
    return version;
  }

  // 公共 API

  /*
   * 上传隐私协议许可
   */
  static Future<bool> updatePrivacyPermissionStatus(bool agree) async {
    return await _channel.invokeMethod("updatePrivacyPermissionStatus",{"status":agree});
  }

  /*
   * 获取 SDK 版本号
   */
  static Future<String> getSDKVersion() async {
    final String ver = await _channel.invokeMethod('getSDKVersion');
    return ver;
  }

  /*
   * 获取regId
   */
  static Future<Map<String, dynamic>> getRegistrationId() async {
    final Map ridMap = await _channel.invokeMethod('getRegistrationId');
    Map<String, dynamic> resMap = Map<String, dynamic>.from(ridMap);
    return resMap;
  }
  
  /*
   * 添加推送回调监听（接收自定义透传消息回调、接收通知消息回调、接收点击通知消息回调、接收别名或标签操作回调）
   */
  static addPushReceiver(EventHandler onEvent, EventHandler onError) {
      _channelReciever.receiveBroadcastStream().listen(onEvent, onError: onError);
  }

  /*
   * 移除推送回调监听
   */
  static Future<void> removePushReceiver() async {
    await _channel.invokeMethod('removePushReceiver');
  }

  /*
   * 停止推送
   */
  static Future<void> stopPush() async {
    await _channel.invokeMethod('stopPush');
  }

  /*
   * 重新打开推送服务
   */
  static Future<void> restartPush() async {
    await _channel.invokeMethod('restartPush');
  }

  /*
   * 是否已停止接收推送
   */
  static Future<bool> isPushStopped() async {
    return await _channel.invokeMethod('isPushStopped'); 
  }

  /*
   * 设置别名
   */
  static Future<Map<String, dynamic>> setAlias (String alias) async {
    final Map aliasMap = await _channel.invokeMethod('setAlias', {"alias": alias});
    Map<String, dynamic> resMap = Map<String, dynamic>.from(aliasMap);
    return resMap;
  }

  /*
   * 获取别名
   */
  static Future<Map<String, dynamic>> getAlias() async {
    final Map aliasMap = await _channel.invokeMethod('getAlias');
    Map<String, dynamic> resMap = Map<String, dynamic>.from(aliasMap);
    return resMap;
  }

  /*
   * 删除别名
   */
  static Future<Map<String, dynamic>> deleteAlias() async {
    final Map aliasMap = await _channel.invokeMethod('deleteAlias');
    Map<String, dynamic> resMap = Map<String, dynamic>.from(aliasMap);
    return resMap;
  }

  /*
   * 添加标签
   */
  static Future<Map<String, dynamic>> addTags (List<String> tags) async {
    final Map tagsMap = await _channel.invokeMethod('addTags', {"tags": tags});
    Map<String, dynamic> resMap = Map<String, dynamic>.from(tagsMap);
    return resMap;
  }

  /*
   * 获取标签
   */
  static Future<Map<String, dynamic>> getTags() async {
    final Map tagsMap = await _channel.invokeMethod('getTags');
    Map<String, dynamic> resMap = Map<String, dynamic>.from(tagsMap);
    return resMap;
  }

  /*
   * 删除标签
   */
  static Future<Map<String, dynamic>> deleteTags (List<String> tags) async {
    final Map tagsMap = await _channel.invokeMethod('deleteTags', {"tags": tags});
    Map<String, dynamic> resMap = Map<String, dynamic>.from(tagsMap);
    return resMap;
  }

  /*
   * 清空标签
   */
  static Future<Map<String, dynamic>> cleanTags() async {
    final Map tagsMap = await _channel.invokeMethod('cleanTags');
    Map<String, dynamic> resMap = Map<String, dynamic>.from(tagsMap);
    return resMap;
  }

  /*
   * 发送本地通知
   */
  static Future<void> addLocalNotification(MobPushLocalNotification localNotification) async {
    await _channel.invokeMethod('addLocalNotification', {"localNotification":json.encode(localNotification.toJson())});
  }

  /*
   * 绑定手机号
   */
  static Future<Map<String, dynamic>> bindPhoneNum (String phoneNum) async {
    final Map phoneMap = await _channel.invokeMethod('bindPhoneNum', {"phoneNum": phoneNum});
    Map<String, dynamic> resMap = Map<String, dynamic>.from(phoneMap);
    return resMap;
  }

  /*
   * 测试模拟推送，用于测试
   * type：模拟消息类型，1、通知测试；2、内推测试；3、定时
   * content：模拟发送内容，500字节以内，UTF-8
   * space：仅对定时消息有效，单位分钟，默认1分钟
   * extras: 附加数据，json字符串
   */
  static Future<Map<String, dynamic>> send(int type, String content, int space, String extras) async {
    final Map sendMap = await _channel.invokeMethod("send", {"type": type, "content": content, "space": space, "extrasMap": extras});
    Map<String, dynamic> resMap = Map<String, dynamic>.from(sendMap);
    return resMap;
  }

  // Android API
  /*
   * 设置点击通知是否跳转默认页(仅andorid)
   */
  static Future<void> setClickNotificationToLaunchMainActivity (bool enable) async {
      await _channel.invokeMethod('setClickNotificationToLaunchMainActivity', {"enable": enable});
  }

  /*
   * 移除本地通知(仅andorid)
   */
  static Future<bool> removeLocalNotification(int notificationId) async {
    final bool result = await _channel.invokeMethod('removeLocalNotification', {"notificationId": notificationId});
    return result;
  }

  /*
   * 清空本地通知(仅andorid)
   */
  static Future<bool> clearLocalNotifications() async {
    final bool result = await _channel.invokeMethod('clearLocalNotifications');
    return result;
  }

  /*
   * 设置通知栏icon，不设置默认取应用icon(仅andorid)
   */
  static Future<void> setNotifyIcon(String resId) async {
   await _channel.invokeMethod('setNotifyIcon',{"iconRes": resId});
  }

  /*
   * 设置应用在前台时是否隐藏通知不进行显示，不设置默认不隐藏通知(仅andorid)
   */
  static Future<void> setAppForegroundHiddenNotification (bool hidden) async {
    await _channel.invokeMethod('setAppForegroundHiddenNotification', {"hidden": hidden});
  }

  /*
   * 设置是否显示角标(仅andorid)
   */
  static Future<void> setShowBadge (bool show) async {
    await _channel.invokeMethod('setShowBadge', {"show": show});
  }

  /*
   * 设置通知静音时段（推送选项）(仅andorid)
   * @param startHour   开始时间[0~23] (小时)
   * @param startMinute 开始时间[0~59]（分钟）
   * @param endHour     结束时间[0~23]（小时）
   * @param endMinute   结束时间[0~59]（分钟）
   */
  static Future<void> setSilenceTime(int startHour, int startMinute, int endHour, int endMinute) async {
   await _channel.invokeMethod('setSilenceTime',
        {"startHour": startHour, "startMinute": startMinute, "endHour": endHour, "endMinute": endMinute});
  }

  // iOS API
  
  /*
   * 设置远程推送，向用户授权(仅 iOS)
   */
  static Future<void> setCustomNotification() async {
    await _channel.invokeMethod('setCustomNotification');
  }
  
  /*
   * 设置远程推送环境 (仅 iOS)
   * @param isPro  开发环境 false, 线上环境 true
   */
  static Future<void> setAPNsForProduction(bool isPro) async {
    await _channel.invokeMethod('setAPNsForProduction', {"isPro": isPro});
  }

  /*
   * 设置角标 (仅 iOS)
   * @param badge  角标数量
   */
  static Future<void> setBadge(int badge) async {
    await _channel.invokeMethod('setBadge', {"badge": badge});
  }

  /*
   * 清空角标，不清除通知栏消息记录 (仅 iOS)
   */
  static Future<void> clearBadge() async {
    await _channel.invokeMethod('clearBadge');
  }

  /*
   * 设置应用在前台有 Badge、Sound、Alert 三种类型,默认3个选项都有,iOS 10以上设置生效.(仅 iOS)
   * type: 0->None , 1->仅Badge, 2->仅Sound, 4->仅Alert, 5->Badge+Alert, 6->Sound+Alert, 7->Badge+Sound+Alert
  */
  static Future<void> setAPNsShowForegroundType(int type) async {
    await _channel.invokeMethod('setAPNsShowForegroundType', {'type': type});
  }

}
